src/java/htsjdk/samtools/CRAMFileReader.java

```diff
@@ -562,64 +495,46 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
 
         @Override
         public void close() {
-            delegate.close();
+            if (unfilteredIterator != null) {
+                unfilteredIterator.close();
+            }
         }
 
         @Override
         public boolean hasNext() {
-            if (next != null) return true;
-            if (noMore) return false;
-
-            while (delegate.hasNext()) {
-                next = delegate.next();
-
-                if (isWithinTheInterval(next)) break;
-                if (isBeyondTheInterval(next)) {
-                    next = null;
-                    noMore = true;
-                    return false;
-                }
-                next = null;
+            if (nextRec != null) {
+                return true;
+            } else if (!hasMore) {
+                return false;
             }
-
-            return next != null;
-        }
-
-        boolean isWithinTheInterval(final SAMRecord record) {
-            final boolean refMatch = record.getReferenceIndex() == interval.referenceIndex;
-            if (interval.start == -1) return refMatch;
-            if (!refMatch) return false;
-
-            final int start = record.getAlignmentStart();
-            final int end = record.getAlignmentEnd();
-            if (overlap) {
-                return CoordMath.overlaps(start, end, interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end);
-            } else {
-                // contained:
-                return CoordMath.encloses(interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end, start, end);
+            while (nextRec == null && hasMore && unfilteredIterator.hasNext()) {
+                SAMRecord nextRecord = unfilteredIterator.next();
+                switch(compareToFilter(nextRecord)) {
+                    case MATCHES_FILTER:
+                        nextRec = nextRecord;
+                        break;
+                    case CONTINUE_ITERATION:
+                        continue;
+                    case STOP_ITERATION:
+                        hasMore = false;
+                        break;
+                    default:
+                        throw new SAMException("Unexpected return from compareToFilter");
+                }
             }
-
-        }
-
-        boolean isBeyondTheInterval(final SAMRecord record) {
-            if (record.getReadUnmappedFlag()) return false;
-            if (record.getReferenceIndex() > interval.referenceIndex) return true;
-            if (record.getReferenceIndex() != interval.referenceIndex) return false;
-
-            return interval.end != -1 && record.getAlignmentStart() > interval.end;
-
+            return nextRec != null;
         }
 
         @Override
         public SAMRecord next() {
-            final SAMRecord result = next;
-            next = null;
+            final SAMRecord result = nextRec;
+            nextRec = null;
```

Comment:
As discussed above, ideally `next()` should advance, and `nextRec == null` would be the termination condition for the iteration.